toolchain/parser/parser_impl.cpp

```diff
@@ -362,6 +368,38 @@ auto ParseTree::Parser::ParsePattern(PatternKind kind) -> llvm::Optional<Node> {
                    /*has_error=*/!type);
   }
 
+  auto valid_me_param =
+      (kind == PatternKind::MeParameter && NextTokenIs(TokenKind::Me()) &&
+       tokens_.GetKind(*(position_ + 1)) == TokenKind::Colon() &&
+       tokens_.GetKind(*(position_ + 2)) == TokenKind::Self());
```

Comment:
There's no requirement that `Self` be the type specified in a `me` binding nor that `Self*` be the type specified in an `addr me` binding, and we intend to allow other types to appear in that position. I think we should just be looking for `me:` or `addr me:` here, and then parse a type after that in the same way we do for the above `PatternBinding` case.